File: morkato/ability.py
from __future__ import annotations
from typing_extensions import Self
from .user import UserTypeFlags
from .utils import NoNullDict
from typing import (
  TYPE_CHECKING,
  SupportsInt,
  Optional
)
if TYPE_CHECKING:
  from .types import (
    Ability as AbilityPayload
  )
  from .state import MorkatoConnectionState
  from .guild import Guild
class Ability:
  def __init__(self, state: MorkatoConnectionState, guild: Guild, payload: AbilityPayload) -> None:
    self.state = state
    self.http = state.http
    self.guild = guild
    self.id = int(payload["id"])
    self.from_payload(payload)
  def from_payload(self, payload: AbilityPayload) -> None:
    self.name = payload["name"]
    self.percent = payload["percent"]
    self.user_type = UserTypeFlags(payload["user_type"])
    self.description = payload["description"]
    self.banner = payload["banner"]
  async def update(
    self, *,
    name: Optional[str] = None,
    user_type: Optional[SupportsInt] = None,
    percent: Optional[int] = None,
    description: Optional[str] = None,
    banner: Optional[str] = None
  ) -> Self:
    payload = NoNullDict(
      name=name,
      user_type=user_type,
      percent=percent,
      description=description,
      banner=banner
    )
    if not payload:
      return self
    payload = await self.http.update_ability(self.guild.id, self.id, **payload)
    self.from_payload(payload)
    return self
  async def delete(self) -> Self:
    payload = await self.http.delete_ability(self.guild.id, self.id)
    self.from_payload(payload)
    self.guild.abilities.remove(self)
    return self

File: api/src/main/kotlin/morkato/api/model/ability/Ability.kt
package morkato.api.model.ability

import morkato.api.infra.repository.AbilityRepository
import org.jetbrains.exposed.sql.ResultRow
import morkato.api.model.guild.Guild
import java.math.BigDecimal

class Ability(
  val guild: Guild,
  val id: Long,
  val name: String,
  val percent: BigDecimal,
  val userType: Int,
  val description: String?,
  val banner: String?
) {
  public constructor(guild: Guild, row: ResultRow) : this(guild, AbilityRepository.AbilityPayload(row));
  public constructor(guild: Guild, payload: AbilityRepository.AbilityPayload) : this(
    guild = guild,
    payload.id,
    payload.name,
    payload.percent,
    payload.userType,
    payload.description,
    payload.banner
  );
  fun update(
    name: String?,
    userType: Int?,
    percent: BigDecimal?,
    description: String?,
    banner: String?
  ) : Ability {
    AbilityRepository.updateAbility(
      guildId = this.guild.id,
      id = this.id,
      name = name,
      percent = percent,
      userType = userType,
      description = description,
      banner = banner
    )
    return Ability(
      guild = this.guild,
      id = this.id,
      name = name ?: this.name,
      percent = percent ?: this.percent,
      userType = userType ?: this.userType,
      description = description ?: this.description,
      banner = banner ?: this.banner
    )
  }
  fun delete() : Unit {
    AbilityRepository.deleteAbility(this.guild.id, this.id)
  }
}